File: server/utils/queue/tokenManager.js
import logger from "../logger.js";
import { getTokenListJson, saveTokenListJson } from "../utils.js";

class TokenManager {
    constructor() {
        this.queue = this.#readTokenQueue() || [];
    }

    #readTokenQueue() {
        return getTokenListJson();
    }

    #saveTokenQueue() {
        saveTokenListJson(this.queue);
    }
    isTokenExist(token){
        return this.queue.some(item => item.token === token);
    }

    isDuplicate(token, username) {
        return this.queue.some(item => item.token === token || item.username === username);
    }

    addToken(item) {
        if (typeof item === "object" && item.token && item.username) {
            if (this.isDuplicate(item.token, item.username)) {
                logger.warn(`Duplicate item not added: ${item.token}`);
                throw new Error(`Duplicate username not allowed!`)
            } else {
                this.queue.push(item);
                this.#saveTokenQueue();
            }
        } else {
            console.error("Invalid input. Item must be an object with 'token' and 'username'.");
        }
    }

    removeTokenByIndex(index) {
        index = index - 1;
        if (index >= 0 && index < this.queue.length) {
            const removedItem = this.queue.splice(index, 1)[0];
            this.#saveTokenQueue();
            return removedItem;
        } else {
            console.error("Invalid index or queue is empty.");
            return null;
        }
    }

    printQueue() {
        return this.queue;
    }
}

export default TokenManager;

File: app/managers/token_manager.py
from app.core import logger
from app.core.utils import get_token_list_json, save_token_list_json

class TokenManager:
    def __init__(self):
        self.queue = self._read_token_queue() or []
    
    def _read_token_queue(self):
        return get_token_list_json()
    
    def _save_token_queue(self):
        save_token_list_json(self.queue)
    
    def is_token_exist(self, token):
        return any(item.get('token') == token for item in self.queue)
    
    def is_duplicate(self, token, username):
        return any(
            item.get('token') == token or item.get('username') == username
            for item in self.queue
        )
    
    def add_token(self, item):
        if isinstance(item, dict) and item.get('token') and item.get('username'):
            if self.is_duplicate(item['token'], item['username']):
                logger.warn(f"Duplicate item not added: {item['token']}")
                raise Exception("Duplicate username not allowed!")
            else:
                self.queue.append(item)
                self._save_token_queue()
        else:
            logger.error("Invalid input. Item must be an object with 'token' and 'username'.")
    
    def remove_token_by_index(self, index):
        index = index - 1
        if 0 <= index < len(self.queue):
            removed_item = self.queue.pop(index)
            self._save_token_queue()
            return removed_item
        else:
            logger.error("Invalid index or queue is empty.")
            return None
    
    def print_queue(self):
        return self.queue
